src/commands/SinceCommand.ts

```diff
@@ -0,0 +1,240 @@
+/*
+Copyright 2020 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { Mjolnir } from "../Mjolnir";
+import { LogLevel, LogService, RichReply } from "matrix-bot-sdk";
+import { htmlEscape, parseDuration } from "../utils";
+import { ParseEntry } from "shell-quote";
+import { HumanizeDurationLanguage, HumanizeDuration } from "humanize-duration-ts";
+
+const HUMANIZE_LAG_SERVICE: HumanizeDurationLanguage = new HumanizeDurationLanguage();
+const HUMANIZER: HumanizeDuration = new HumanizeDuration(HUMANIZE_LAG_SERVICE);
+
+enum Action {
+    Kick = "kick",
+    Ban = "ban",
+    Show = "show"
+}
+
+type Result<T> = {ok: T} | {error: string};
+
+function parseToken<T>(name: string, token: ParseEntry, parser: (source: string) => Result<T>): Result<T> {
+    if (!token) {
+        return { error: `Missing ${name}`};
+    }
+    if (typeof token === "object" && "pattern" in token) {
+        // In future versions, we *might* be smarter patterns, but not yet.
+        token = token.pattern;
+    }
+
+    if (typeof token !== "string") {
+        return { error: `Invalid ${name}` };
+    }
+    const result = parser(token);
+    if ("error" in result) {
+        if (result.error) {
+            return { error: `Invalid ${name} ${htmlEscape(token)}: ${result.error}`};
+        } else {
+            return { error: `Invalid ${name} ${htmlEscape(token)}`};
+        }
+    }
+    return result;
+}
+
+function getTokenAsString(name: string, token: ParseEntry): {error: string}|{ok: string} {
+    if (!token) {
+        return { error: `Missing ${name}`};
+    }
+    if (typeof token === "object" && "pattern" in token) {
+        // In future versions, we *might* be smarter patterns, but not yet.
+        token = token.pattern;
+    }
+    if (typeof token === "string") {
+        return {ok: token};
+    }
+    return { error: `Invalid ${name}` };
+}
+
+// !mjolnir since <date>/<duration> <action> <number> [reason] [...rooms]
```

Comment:
reason can only be one word?

Reply:
The reason I'm using the shell parser is so that we can write `"a complex reason"` (with quotes). But you're right that this doesn't match usual Mjölnir conventions.